pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/OpChainSchedulerService.java

```diff
@@ -31,151 +32,73 @@
+          thrown = e;
+        } finally {
+          _submittedOpChainMap.remove(operatorChain.getId());
+          if (returnedErrorBlock != null || thrown != null) {
+            if (thrown == null) {
+              thrown = new RuntimeException("Error block " + returnedErrorBlock.getDataBlock().getExceptions());
             }
+            operatorChain.cancel(thrown);
+          } else if (isFinished) {
+            operatorChain.close();
           }
         }
-      });
-    }
-  }
-
-  /**
-   * Register a new operator chain with the scheduler.
-   *
-   * @param operatorChain the chain to register
-   */
-  public final void register(OpChain operatorChain) {
-    operatorChain.getStats().queued();
-    _scheduler.register(operatorChain);
-    LOGGER.debug("({}): Scheduler is now handling operator chain listening to mailboxes {}. "
-            + "There are a total of {} chains awaiting execution.", operatorChain,
-        operatorChain.getReceivingMailboxIds(),
-        _scheduler.size());
-  }
-
-  /**
-   * Async cancel a request. Request will not be fully cancelled until the next time opChain is being polled.
-   *
-   * @param requestId requestId to be cancelled.
-   */
-  public final void cancel(long requestId) {
-    _cancelledRequests.put(requestId, requestId);
-  }
-
-  /**
-   * This method should be called whenever data is available for an {@link OpChain} to consume.
-   * Implementations of this method should be idempotent, it may be called in the scenario that no data is available.
-   *
-   * @param opChainId the identifier of the operator chain
-   */
-  public final void onDataAvailable(OpChainId opChainId) {
-    _scheduler.onDataAvailable(opChainId);
-  }
-
-  // TODO: remove this method after we pipe down the proper executor pool to the v1 engine
-  public ExecutorService getWorkerPool() {
-    return _workerPool;
-  }
-
-  private void closeOpChain(OpChain opChain) {
-    try {
-      opChain.close();
-    } finally {
-      _scheduler.deregister(opChain);
-    }
+      }
+    });
+    _submittedOpChainMap.put(operatorChain.getId(), scheduledFuture);
   }
 
-  private void cancelOpChain(OpChain opChain, Throwable e) {
-    try {
-      opChain.cancel(e);
-    } finally {
-      _scheduler.deregister(opChain);
+  @Override
+  public void cancel(long requestId) {
+    // simple cancellation. for leaf stage this cannot be a dangling opchain b/c they will eventually be cleared up
+    // via query timeout.
+    List<Map.Entry<OpChainId, Future<?>>> entries =
+        _submittedOpChainMap.entrySet().stream().filter(entry -> entry.getKey().getRequestId() == requestId)
+            .collect(Collectors.toList());
+    for (Map.Entry<OpChainId, Future<?>> entry : entries) {
+      OpChainId key = entry.getKey();
+      Future<?> future = entry.getValue();
+      if (future != null) {
+        future.cancel(true);
+      }
+      _submittedOpChainMap.remove(key);
```

Comment:
I didn't want to use that approach because Map.entrySet says:

```If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation, or through the setValue operation on a map entry returned by the iterator) the results of the iteration are undefined```

On the other hand, ConcurrentHashMap.entrySet says:
```
The view's iterators and spliterators are [weakly consistent](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly).
```
Which is defined in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly as:
```
Most concurrent Collection implementations (including most Queues) also differ from the usual java.util conventions in that their [Iterators](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html) and [Spliterators](https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html) provide weakly consistent rather than fast-fail traversal:

- they may proceed concurrently with other operations
- they will never throw [ConcurrentModificationException](https://docs.oracle.com/javase/8/docs/api/java/util/ConcurrentModificationException.html)
- they are guaranteed to traverse elements as they existed upon construction exactly once, and may (but are not guaranteed to) reflect any modifications subsequent to construction.
```

So your code should be safe, but seems difficult to reason about that. We also don't know the blocking implications, although they doesn't seem to be that important (we assume they are correct, but we don't care that much if we block other accesses to the map for some extra microseconds).

That is why I though the current code is easier to reason about, but I'm open to use yours if you prefer that.

Reply:
I actually have the opposite opinion here. With `iterator` I'm pretty sure I can get the intended behavior assuming `ConcurrentHashMap` is a very well tested class. If I first cache all the entries into a separate list, and then iterate over them without `ConcurrentHashMap` knowing this list's existence, I cannot guarantee the map won't modify the entry. So IMO it is actually safer to use the `iterator`